coalib/io/FileCache.py

```diff
@@ -0,0 +1,137 @@
+import functools
+import sys
+import os
+import tokenize
+
+__all__ = ["get_content", "clearcache", "checkcache"]
+
+
+# The cache. Maps filenames to either a thunk which will provide source code,
+# or a tuple (size, mtime, contents, fullname) once loaded.
+cache = {}
+
+
+def clearcache():
+    """
+    Clear the cache entirely.
+    """
+    global cache
+    cache = {}
+
+
+def get_content(filename, module_globals=None):
+    """
+    Get the raw contents of a file from the cache.
+    Update the cache if it doesn't contain an entry for this file already.
+    """
+    if filename in cache:
+        entry = cache[filename]
+        if len(entry) != 1:
+            return cache[filename][2]
+
+    try:
+        return updatecache(filename, module_globals)
+    except MemoryError:
+        clearcache()
+        return None
+
+
+def checkcache(filename=None):
+    """
+    Discard cache entries that are out of date.
+    (This is not checked upon each call!)
+    """
+
+    if filename is None:
+        filenames = list(cache.keys())
+    else:
+        if filename in cache:
+            filenames = [filename]
+        else:
+            return
+
+    for filename in filenames:
+        entry = cache[filename]
+        if len(entry) == 1:
+            # lazy cache entry, leave it lazy.
+            continue
+        size, mtime, lines, fullname = entry
+        if mtime is None:
+            continue   # no-op for files loaded via a __loader__
+        try:
+            stat = os.stat(fullname)
+        except OSError:
+            del cache[filename]
+            continue
+        if size != stat.st_size or mtime != stat.st_mtime:
+            del cache[filename]
+
+
+def updatecache(filename, module_globals=None):
+    """
+    Update a cache entry and return its list of lines.
+    If something's wrong, print a message, discard the cache entry,
+    and return an empty list.
+    """
+
+    if filename in cache:
+        if len(cache[filename]) != 1:
+            del cache[filename]
+    if not filename or (filename.startswith('<') and filename.endswith('>')):
+        return None
+
+    fullname = filename
+    try:
+        stat = os.stat(fullname)
+    except OSError:
+        basename = filename
+
+        # Realise a lazy loader based lookup if there is one
+        # otherwise try to lookup right now.
+        if lazycache(filename, module_globals):
+            try:
+                data = cache[filename][0]()
+            except (ImportError, OSError):
+                pass
+            else:
+                if data is None:
+                    # No luck, the PEP302 loader cannot find the source
+                    # for this module.
+                    return None
+                cache[filename] = (
+                    len(data), None,
+                    data, fullname
+                )
+                return cache[filename][2]
+
+    try:
+        with open(fullname, 'rb') as fp:
```

Comment:
This seemed to have done the trick. Now instead of a list the file contents are stored as bytes.